samples/packages/lake-formation-tag-sync/src/main/kotlin/com/atlan/pkg/lftag/CSVProducer.kt

```diff
@@ -0,0 +1,80 @@
+    private val connectionMap: Map<String, String>,
+    private val metadataMap: Map<String, String>,
+    private val outputDirName: String,
+) {
+    private val logger = KotlinLogging.logger {}
+    private val mapper = jacksonObjectMapper()
+    private val missingConnectionKeys = mutableSetOf<String>()
+    private val missingTagKeys = mutableSetOf<String>()
+    private val headerNames = setOf(NAME, TYPE_NAME, QUALIFIED_NAME) + (metadataMap.map { (_, value) -> value }).toSet()
+
+    fun transform(inputFileName: String, outputFileName: String) {
+        val fileName = "${this.outputDirName}${File.separator}$outputFileName"
+
+        CSVWriter(fileName).use { csv ->
+            csv.writeHeader(headerNames)
+            val start = System.currentTimeMillis()
+            val jsonString: String = File(inputFileName).readText(Charsets.UTF_8)
+            val tagData = mapper.readValue(jsonString, TagData::class.java)
+            tagData.tableList.forEach { tableInfo ->
+                val table = tableInfo.table
+                val (connectionKey, schemaName) = table.databaseName.split('_', limit = 2)
+                if (connectionKey in this.connectionMap) {
+                    val qualifiedName = "${connectionMap.getValue(connectionKey)}/$schemaName/${table.name}"
+                    var row = mutableMapOf(QUALIFIED_NAME to qualifiedName, TYPE_NAME to "Table", NAME to table.name)
+                    getTagValues(tableInfo.lfTagsOnTable, row)
+                    csv.writeHeader(rowAsList(row))
+                    tableInfo.lfTagsOnColumn.forEach { column ->
+                        val columnQualifiedName = "$qualifiedName/${column.name}"
+                        row = mutableMapOf(
+                            QUALIFIED_NAME to columnQualifiedName,
+                            TYPE_NAME to "Column",
+                            NAME to table.name,
+                        )
+                        getTagValues(column.lfTags, row)
+                        csv.writeHeader(rowAsList(row))
+                    }
+                } else {
+                    missingConnectionKeys.add(connectionKey)
+                }
+            }
+            logger.info { "Total time taken: ${System.currentTimeMillis() - start} ms" }
+        }
+    }
+
+    private fun getTagValues(tags: List<LFTagPair>, row: MutableMap<String, String>) {
+        tags.forEach { tag ->
+            val tagKey = if (tag.tagKey.startsWith("subdomain")) "Subdomain" else tag.tagKey
+            if (tagKey in this.metadataMap) {
+                val cmName = this.metadataMap.getValue(tagKey)
+                row[cmName] = tag.tagValues[0]
+            }
+            this.missingTagKeys.add(tagKey)
+        }
+    }
+
+    private fun rowAsList(row: Map<String, String>): List<String> {
+        val output = mutableListOf<String>()
+        headerNames.forEach { header ->
+            output.add(row.getOrDefault(header, ""))
+        }
+        return output
+    }
```

Comment:
I think there's now a `writeRecord` method in the `CSVWriter` class which will do this translation for you already (from a map keyed by column name)